### PR TITLE
fix #4312: class based api

### LINKED ISSUE #4312: move to class based api
Move all api methods in flask into a class. This is part of adding a request context to all apis (https://github.com/radiasoft/sirepo/issues/4234)

- there will be a base class for APIs
- every module that contains an API will have its own class that inherits from the base
- write a program that changes our code to do this
- call_api will have to be handled by hand to generate the object

### PR DESCRIPTION
**Pull request by rorour**
@robnagler Currently the API class is being instantiated in uri_router.register_api_module() - not sure if that's really the correct place, but with how things are set up right now it seemed necessary, since each module has its own API class which needs to be instantiated in order to call the apis.

Diffs for the apis might look confusing. The only changes I made were to indent them inside of the API class and add `self` as a parameter.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,7 +5,7 @@
   "devDependencies": {
     "jasmine-core": "latest",
     "jshint": "latest",
-    "karma": "latest",
+    "karma": "^6.3.19",
```
Comment (robnagler) [inline package.json:8]: I think we want to push this change in a separate PR. Please create an issue and move that change there.

File: sirepo/api.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+class APIRequest:
```
Comment (robnagler) [inline sirepo/api.py:1]: Add our standard file header.

Also document the class. It's holds the request context for all API calls. On reading this, I think we should rename this module to sirepo.request and the class would be called Base. I comment on how to automate these changes.

File: sirepo/api.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    def __init__(self):
```
Comment (robnagler) [inline sirepo/api.py:2]: If there's nothing in `__init__` then don't have one so that the superclass init is automatically called.

File: sirepo/api.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+        return
+
+    def call_api(self, name, kwargs=None, data=None):
```
Comment (robnagler) [inline sirepo/api.py:5]: Write some documentation describing this call. It will be used to call other APIs, checks permissions, etc.

Also add some documentation to uri_router.call_api that it should not be called, and instead self.call_api should be used.

File: sirepo/api.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+        return uri_router.call_api(name, kwargs=kwargs, data=data)
+
+
+APIBase = APIRequest
```
Comment (robnagler) [inline sirepo/api.py:9]: This was a quick fix. Please change all sirepo.api.APIBase to sirepo.request.Base and remove this line. I tested this so it should work:

```sh
cd ~/src/radiasoft/sirepo/sirepo
perl  -pi -e 's{(?<=sirepo.)api.APIRequest}{request.Base}; s{(?<=import sirepo.)api$}{request}; s{(?<=class )API\b}{Request}' *.py */*.py
```

Run this on a committed so that you can verify/undo these changes.

File: sirepo/uri_router.py [hunk 6]
```diff
@@ -190,6 +203,22 @@ def uri_for_api(api_name, params=None, external=True):
     return res
 
 
+class _Route(PKDict):
```
Comment (robnagler) [inline sirepo/uri_router.py:206]: Document this: holds all route information. Document the known keys: func, cls, name, func_name, etc.

File: sirepo/uri_router.py [hunk 7]
```diff
@@ -157,11 +167,14 @@ def register_api_module(module=None):
     if hasattr(m, 'init_apis'):
         m.init_apis()
     # It's ok if there are no APIs
-    for n, o in inspect.getmembers(m):
-        if n.startswith(_FUNC_PREFIX) and inspect.isfunction(o):
+    if not hasattr(m, 'API'):
```
Comment (robnagler) [inline sirepo/uri_router.py:170]: Document why this test exists. It doesn't seem like the right place. Probably should be in `_register_sim_api_module`.

File: sirepo/uri_router.py [hunk 8]
```diff
@@ -157,11 +167,14 @@ def register_api_module(module=None):
...
+        return
+    c = m.API
```
Comment (robnagler) [inline sirepo/uri_router.py:172]: change to `m.Request`.

File: sirepo/uri_router.py [hunk 9]
```diff
@@ -157,11 +167,14 @@ def register_api_module(module=None):
...
+    for n, o in inspect.getmembers(c):
+        if n.startswith(_FUNC_PREFIX) and inspect.isfunction(o) and n in c.__dict__:
```
Comment (robnagler) [inline sirepo/uri_router.py:174]: move this condition to a separate function that is named something appropriate.

**Review by robnagler — CHANGES_REQUESTED** [review]
Looks good. Needs more documentation though.

File: sirepo/uri_router.py (hunk 7 above)
Comment (rorour) [reply sirepo/uri_router.py:170]: @robnagler Without this check, tests fail because some modules like sirepo.auth.basic don't have any APIs and therefore don't have the class. Could you expand on why `_register_sim_api_modules` is the correct place and how exactly to add it there?

Comment (robnagler) [reply sirepo/uri_router.py:170]: Ah, I didn't remember that. I was thinking it was only sim api modules that didn't have APIs. Leave as is, then, and document including an example module.

Technically, this is an abuse of uri_router to enforce initialization. We really should sirepo.init (or something) that modules register with and declare dependencies. I thought [about this 7 years ago](https://github.com/radiasoft/pykern/wiki/ImportThoughts) (almost to the day), but never implemented. The ad hoc initialization that we are doing now is not working, because it requires too much special knowledge/handling. I'll add an issue.

File: sirepo/uri_router.py (hunk 9 above)
Comment (rorour) [reply sirepo/uri_router.py]: 3ccfbc6e956420f7da6fc23176cdebb7ae57cfc9

File: sirepo/uri_router.py (hunk 8 above)
Comment (rorour) [reply sirepo/uri_router.py]: 3ccfbc6e956420f7da6fc23176cdebb7ae57cfc9

File: sirepo/uri_router.py (hunk 7 above)
Comment (rorour) [reply sirepo/uri_router.py]: 3ccfbc6e956420f7da6fc23176cdebb7ae57cfc9

File: sirepo/uri_router.py (hunk 6 above)
Comment (rorour) [reply sirepo/uri_router.py:211]: 3ccfbc6e956420f7da6fc23176cdebb7ae57cfc9

File: sirepo/api.py (hunk 5 above)
Comment (rorour) [reply sirepo/api.py]: 3ccfbc6e956420f7da6fc23176cdebb7ae57cfc9

File: sirepo/api.py (hunk 4 above)
Comment (rorour) [reply sirepo/api.py]: 3ccfbc6e956420f7da6fc23176cdebb7ae57cfc9

File: sirepo/api.py (hunk 2 above)
Comment (rorour) [reply sirepo/api.py]: 3ccfbc6e956420f7da6fc23176cdebb7ae57cfc9

File: sirepo/api.py (hunk 3 above)
Comment (rorour) [reply sirepo/api.py]: 3ccfbc6e956420f7da6fc23176cdebb7ae57cfc9

File: package.json (hunk 1 above)
Comment (rorour) [reply package.json]: https://github.com/radiasoft/sirepo/issues/4413

File: sirepo/uri_router.py [hunk 10]
```diff
@@ -204,6 +209,13 @@ def uri_for_api(api_name, params=None, external=True):
 
 
 class _Route(PKDict):
+    """ Holds all route information for an API.
+
+    Keys:
+        func (function): The function object for the API
+        cls (class): The class in the API's module that contains the API function.
+        func_name (str): The function's name (ex: 'api_admJobs').
```
Comment (robnagler) [inline sirepo/uri_router.py:217]: Great. Please add more doc for decl_uri, etc. Look in _init_uris to see all the fields.

**Review by robnagler — CHANGES_REQUESTED** [review]
Looks great. Minor doc change requested.

File: sirepo/uri_router.py (hunk 10 above)
Comment (rorour) [reply sirepo/uri_router.py]: 8f39fbdf6e35e7cd3105d1e0fbf9da2e60f893c7

**Review by robnagler — APPROVED** [review]
Great!

**Review by e-carlin — COMMENTED** [review]
Looks awesome! Some minor nitpicks you can address when you have time.

File: sirepo/auth/__init__.py [hunk 11]
```diff
@@ -21,6 +20,7 @@
 import contextlib
 import datetime
 import importlib
+import sirepo.request
```
[the remark this answers is not in the document]
Comment (rorour) [reply sirepo/auth/__init__.py:23]: d49c4cd97a3c14d1d445f78b3e3c5ab78269589b

File: sirepo/uri_router.py [hunk 12]
```diff
@@ -70,19 +88,22 @@ def call_api(func_or_name, kwargs=None, data=None):
...
             s = sirepo.http_request.set_sim_type(kwargs.get('simulation_type'))
-        f = check_api_call(func_or_name)
+        else:
+            kwargs = PKDict()
+        f = _check_api_call(route_or_name)
         try:
             if data:
                 p = sirepo.http_request.set_post(data)
-            r = flask.make_response(f(**kwargs) if kwargs else f())
+            i = f.cls()
```
[the remark this answers is not in the document]
Comment (rorour) [reply sirepo/uri_router.py:97]: b562769b31804e951044ff01bebafaaeb1c17434

File: sirepo/uri_router.py [hunk 13]
```diff
@@ -50,13 +51,30 @@
...
+    """Check if `name` is executable and in allowed
+
+    Args:
+        name (str): name of the api
+        allowed (tuple): names that are allowed to be called
+    Returns:
+        str: api name
+    """
+    _check_api_call(name)
+    assert allowed, 'must pass non-zero set'
```
[the remark this answers is not in the document]
Comment (rorour) [reply sirepo/uri_router.py:64]: e676a795bb22b1e6c52c63a3458e39e512a0258b

File: sirepo/uri_router.py [hunk 14]
```diff
@@ -50,13 +51,30 @@
...
+    if name not in allowed:
+        raise AssertionError(f'api={name} not in allowed={allowed}')
+
+
+def call_api(route_or_name, kwargs=None, data=None):
+    """Should not be called outside of Base.call_api(). Use self.call_api() to call API.
```
[the remark this answers is not in the document]
Comment (rorour) [reply sirepo/uri_router.py:70]: ec472d9bbdb74b05d0d0cebc3f0f3b039f997c79
